code/modules/mob/living/living_healthscan.dm

```diff
@@ -0,0 +1,655 @@
+/*
+* Health scanning of living mobs
+* Uses a datum like player panels
+* Tgui mostly
+* can be done by a variety of sources - health scanners, body scanners, ghost verb
+*/
+
+GLOBAL_LIST_INIT(known_implants, subtypesof(/obj/item/implant))
+
+/// var reffing this on /mob/living is called HEALTH_DISPLAY
+/datum/health_scan
+	var/mob/living/target_mob
+	var/detail_level = DETAIL_LEVEL_FULL
+
+/datum/health_scan/New(var/mob/target)
+	. = ..()
+	target_mob = target
+
+/datum/health_scan/Destroy(force, ...)
+	target_mob = null
+	SStgui.close_uis(src)
+	return ..()
+
+/// This is the proc for interacting with, or looking at, a mob's health display. Also contains skillchecks and the like. You may NOT call tgui interact directly, and you MUST set the detail level.
+/datum/health_scan/proc/look_at(mob/user, var/detail = DETAIL_LEVEL_FULL, var/bypass_checks = FALSE, var/ignore_delay = TRUE, var/alien = FALSE, datum/tgui/ui)
+	if(!bypass_checks)
+		if(HAS_TRAIT(target_mob, TRAIT_FOREIGN_BIO) && !alien)
+			to_chat(user, SPAN_WARNING("ERROR: Unknown biology detected."))
+			return
+		if(!(ishuman(user) || SSticker?.mode.name == "monkey"))
+			to_chat(usr, SPAN_WARNING("You don't have the dexterity to do this!"))
+			return
+		if(!ignore_delay && !skillcheck(user, SKILL_MEDICAL, SKILL_MEDICAL_MEDIC))
+			to_chat(user, SPAN_WARNING("You start fumbling around with [target_mob]..."))
+			var/fduration = 60
+			if(skillcheck(user, SKILL_MEDICAL, SKILL_MEDICAL_DEFAULT))
+				fduration = 30
+			if(!do_after(user, fduration, INTERRUPT_NO_NEEDHAND, BUSY_ICON_FRIENDLY) || !user.Adjacent(target_mob))
+				return
+		if(!istype(target_mob, /mob/living/carbon) || isXeno(target_mob))
+			to_chat(user, SPAN_WARNING("The scanner can't make sense of this creature."))
+			return
+
+	detail_level = detail
+	tgui_interact(user, ui)
+
+/datum/health_scan/ui_state(mob/user)
+	if(isobserver(user))
+		return GLOB.always_state
+	else
+		return GLOB.not_incapacitated_state
+
+/datum/health_scan/tgui_interact(mob/user, datum/tgui/ui)
+	if(!target_mob)
+		return
+
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "HealthScan", "[target_mob.name]'s health scan")
+		ui.open()
+		ui.set_autoupdate(FALSE)
+
+
+/datum/health_scan/ui_static_data(mob/user, var/data_detail_level = null)
```

Comment:
I just wonder if it would be better to have different procs to handle different unit types. This is a massive function with quite a lot of parallel logic for synths and humans

Reply:
eh, about 75% is for both synths and humans